### PR TITLE
[TS] LPS-161481 StaleObjectStateException is caused by Hibernate auto-flush behaviour

### PR DESCRIPTION
**Pull request by istvansajtos**
Dear Team,

This is the continuation of https://github.com/liferay-core-infra/liferay-portal-ee/pull/15, as requested, sent to master and with the suggested modifications.

Regarding the simplification of the following statement in updateStatus, it seems safe because the current use cases do not intend to approve already approved users.
```
		if ((status == WorkflowConstants.STATUS_APPROVED) &&
			(user.getStatus() != WorkflowConstants.STATUS_APPROVED))
```

updateStatus() is used via UserService at the following places, the simplification doesn't affect them. 
- [EditAccountUsersMVCActionCommand._updateAccountUsersStatus](https://github.com/liferay/liferay-portal/blob/master/modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/portlet/action/EditAccountUsersMVCActionCommand.java#L100) - NOT RELEVANT - used only for DEACTIVATING or RESTORING account users
- [EditUserMVCActionCommand.deleteUsers](https://github.com/liferay/liferay-portal/blob/master/modules/apps/users-admin/users-admin-web/src/main/java/com/liferay/users/admin/web/internal/portlet/action/EditUserMVCActionCommand.java#L112) - NOT RELEVANT

updateStatus() is used via UserLocalService at the following places, the simplification doesn't affect them. 
- [DocumentLibraryDDMFormFieldTemplateContextContributor._createDDMFormDefaultUser(long companyId)](https://github.com/liferay/liferay-portal/blob/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/java/com/liferay/dynamic/data/mapping/form/field/type/internal/document/library/DocumentLibraryDDMFormFieldTemplateContextContributor.java#L179) - sets INACTIVE status
- [CreateAnonymousAccountMVCActionCommand._addAnonymousUser(ActionRequest actionRequest) ](https://github.com/liferay/liferay-portal/blob/master/modules/apps/login/login-web/src/main/java/com/liferay/login/web/internal/portlet/action/CreateAnonymousAccountMVCActionCommand.java#L265)- sets INCOMPLETE status
- [LDAPUserImporterImpl._updateUser()](https://github.com/liferay/liferay-portal/blob/master/modules/apps/portal-security/portal-security-ldap-impl/src/main/java/com/liferay/portal/security/ldap/internal/exportimport/LDAPUserImporterImpl.java#L1879) - this one has its own check to avoid unnecessary updateStatus call
- [UADAnonymousUserProviderImpl._createAnonymousUser()](https://github.com/liferay/liferay-portal/blob/master/modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/anonymizer/UADAnonymousUserProviderImpl.java#L126) - sets INACTIVE status
- [ErasePersonalDataMVCActionCommand.doTransactionalCommand()](https://github.com/liferay/liferay-portal/blob/master/modules/apps/user-associated-data/user-associated-data-web/src/main/java/com/liferay/user/associated/data/web/internal/portlet/action/ErasePersonalDataMVCActionCommand.java#L66) - sets INACTIVE status

Please let me know if you have any observations.

Thank you,
Istvan

cc @yuhai

### CODE REVIEW

**Comment by Kvale1733** [comment]
ci:test:users

**Comment by Kvale1733** [comment]
@istvansajtos No regressions. Testing passed following ticket steps.

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
 import com.liferay.portal.kernel.model.Role;
 import com.liferay.portal.kernel.model.Ticket;
 import com.liferay.portal.kernel.model.User;
+import com.liferay.portal.kernel.model.UserConstants;
 import com.liferay.portal.kernel.model.UserGroup;
```
Comment (yuhai) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:36]: @istvansajtos, please merge this to the previous commit.

File: portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java [hunk 2]
```diff
@@ -1341,6 +1341,7 @@ public User addUserWithWorkflow(
 
 		workflowServiceContext.setAttribute("autoPassword", autoPassword);
 		workflowServiceContext.setAttribute("sendEmail", sendEmail);
+		workflowServiceContext.setCompanyId(user.getCompanyId());
 
 		user = WorkflowHandlerRegistryUtil.startWorkflowInstance(
```
Comment (yuhai) [inline portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java:1346]: @istvansajtos, I think you consider the case when serviceContext = null in the UserLocalServiceImpl#addUserWithWorkflow(). But the way does not cover the all cases. updateIncompleteUser() method also invoked this. I think we need to place the logic to the common place. 

Refer to WorkflowHandlerRegistryUtil.startWorkflowInstance(), the map workflowContext always include WorkflowConstants.CONTEXT_COMPANY_ID. So we may set this in UserWorkflowHandler#updateStatus().
```
if (serviceContext.getCompanyId() == 0) {
			serviceContext.setCompanyId(
				GetterUtil.getLong(workflowContext.get(
					WorkflowConstants.CONTEXT_COMPANY_ID)));
		}
```

Please also add serviceContext = null test case to cover the fix. You may extract the [codes](https://github.com/liferay-core-infra/liferay-portal/blob/b4cb2aad52beeec74497b0740def8e4ce7319423/modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java#L137-L145) to the private method, named _testAddUser(..., ServiceContext serviceContext), and then use serviceContext = null and serviceContext != null to create two users.

**Comment by yuhai** [comment]
@istvansajtos, please see the above comments and resend it. Thanks.

**Pull request closed without merging**